src/Args.hpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (C) 2020 Joel Rosdahl and other contributors
+//
+// See doc/AUTHORS.adoc for a complete list of contributors.
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#pragma once
+
+#include "system.hpp"
+
+#include "NonCopyable.hpp"
+#include "Util.hpp"
+
+#include "third_party/nonstd/optional.hpp"
+#include "third_party/nonstd/string_view.hpp"
+
+#include <deque>
+#include <string>
+
+class Args
+{
+public:
+  Args();
+  Args(const Args& other);
+
+  static Args from_argv(int argc, const char* const* argv);
+  static Args from_string(const std::string& command);
+  static nonstd::optional<Args> from_gcc_atfile(const std::string& filename);
+
+  Args& operator=(const Args& other);
+
+  bool operator==(const Args& other) const;
+  bool operator!=(const Args& other) const;
+
+  size_t size() const;
+  const std::string& operator[](size_t i) const;
+  std::string& operator[](size_t i);
+
+  // Accessor functions for the legacy API:
+  Args& operator*();
+  const Args* operator->() const;
+
+  class ArgvKeeper : NonCopyable
```

Comment:
We started discussing in [#513](https://github.com/ccache/ccache/pull/513#issuecomment-578486206) but haven't quite finished.
I think this can be simplified **a lot** by having to_argv return the c_str() values of the strings instead of this extra struct. Runtime is also improved by avoiding all those string copies and memory allocations.

It's _slightly_ more complex for usage since to_args return value does no longer outlive Args. But I don't think this is an issue. We are accustomed to such stuff with strings, as in c_str().

Reply:
You could return a vector<char*> to avoid memory management completely.
Call sites would potentially store this locally and use &...[0] to get the char**.

Reply:
> I think this can be simplified a lot by having to_argv return the c_str() values of the strings instead of this extra struct.

In the big scheme of things I don't think that such a simple thing can be simplified _a lot_ – then it would have negative complexity. :slightly_smiling_face: But sure, sounds good. Especially after 9131b0595242 since it wouldn't feel quite right not to copy the string data before that commit.

> You could return a vector<char*> to avoid memory management completely.

One reason I saw for not using a vector is that it has to store the terminating null entry, so `args.size()` will then not be equal to `args.to_argv().size()`, which I think feels slightly strange – it would be better if there were no size method at all. But that's a minor thing; I'm OK with a vector.